File: clients/python/play.py
from time import sleep
import json
import os
import re
import sys
import webbrowser
from requests import RequestException

from game.game_loop import update_game
from game.api import API
from game.models import Game
from ai import AI


UpdatesPerSecond = 3  # how many times we should contact the server per second


def main():
    create_private = "--create-private" in sys.argv or "-c" in sys.argv
    join_private = "--join-private" in sys.argv or "-j" in sys.argv

    player_id, player_secret, api_url = read_config()
    game_id = player_id if join_private else Game.RANKED_GAME_ID
    api = API(player_id, player_secret, api_url)
    previous_tick = -1
    ai = AI()

    if(create_private):
        game_id = api.create_private()
        # This is useful since the game creation is not instant
        sleep(0.5)
        open_browser(game_id, api_url)

    while True:
        try:
            game = api.fetch_game_state(game_id)

            if(game.tick < previous_tick):  # After a game reset, it reinstanciates the AI object
                ai = AI()
            previous_tick = game.tick
            update_game(api, game, ai.step)
            sleep(1 / UpdatesPerSecond)
        except RequestException as e:
            print("You received an network error: %s" % str(e))
            sleep(5)


def read_config():
    ConfigFile = "player.json"
    DefaultConfigFile = "player.default.json"
    DefaultConfigValue = "REPLACEME"  # default value in the config file

    try:
        with open(ConfigFile) as f:
            data = json.load(f)
            id_ =     os.getenv('PLAYER_ID', data["player_id"])
            secret =  os.getenv('PLAYER_SECRET', data["player_secret"])
            api_url = os.getenv('API_URL', data["api_url"])

            if id_ == DefaultConfigValue or secret == DefaultConfigValue or api_url == DefaultConfigValue:
                print("WARNING: Did you forget to change your player "
                      " id/secret/api_url in '%s'?" % ConfigFile, file=sys.stderr)

            return int(id_), secret, api_url

    except FileNotFoundError:
        print("ERROR: Could not find '%s'. "
              "Did you forget to rename '%s' to '%s'?" %
              (ConfigFile, DefaultConfigFile, ConfigFile), file=sys.stderr)
        exit(1)


def open_browser(game_id, api_url):
    webbrowser.open(f"{api_url}/web/index.html?gameId={game_id}", new=2)


if __name__ == "__main__":
    main()

File: clients/java/src/main/java/io/aigar/Play.java
package io.aigar;

import io.aigar.config.Config;
import io.aigar.config.ConfigReader;
import io.aigar.game.Api;
import io.aigar.game.NetworkException;
import io.aigar.game.models.Cell;
import io.aigar.game.models.GameState;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.awt.*;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.Optional;
import java.util.stream.Collectors;

public class Play {
    private static final String BASE_GAME_URL = "web/index.html";
    private static final int RANKED_GAME_ID = -1;
    private static final int UPDATES_PER_SECOND = 3; // how many times we should contact the server per second
    private static final Logger logger = LoggerFactory.getLogger(Play.class);

    public static void main(String[] args) throws IOException, InterruptedException, NetworkException, URISyntaxException {
        // Read input parameters
        boolean createPrivate = false;
        boolean joinPrivate = false;

        for (String argument : args) {
            if (argument.equals("--create-private") || argument.equals("-c"))
                createPrivate = true;
            if (argument.equals("--join-private") || argument.equals("-j"))
                joinPrivate = true;
        }

        // Read config
        ConfigReader configReader = new ConfigReader();
        Config config = configReader.readConfig();

        // Setup game
        Api api = new Api(config.getPlayerId(), config.getPlayerSecret(), config.getApiUrl());
        Ai ai = new Ai();
        int previousTick = -1;

        int gameId = RANKED_GAME_ID;
        if (joinPrivate) {
            gameId = config.getPlayerId();
            logger.info("Joining private game, id: {}", gameId);
        }

        if (createPrivate) {
            gameId = api.createPrivate().getId();
            logger.info("Private game created, id: {}", gameId);

            // This is useful since the game creation is not instant
            Thread.sleep(500);
        }

        if (Desktop.isDesktopSupported())
            Desktop.getDesktop().browse(new URI(config.getApiUrl().concat(BASE_GAME_URL.concat("?gameId=" + gameId))));

        // Game loop
        while (true) {
            GameState gameState = api.fetchGameState(gameId);

            // After a game reset, it reinstanciates the AI object
            if (gameState.getTick() < previousTick)
                ai = new Ai();
            previousTick = gameState.getTick();

            // Actual loop
            ai.step(gameState);
            api.sendActions(gameId, gameState.getMe()
                    .getCells()
                    .stream()
                    .map(Cell::getActions)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .collect(Collectors.toList()));

            Thread.sleep(1000 / UPDATES_PER_SECOND);
        }
    }
}
